compiler/rustc_mir_build/src/build/scope.rs

```diff
@@ -1094,21 +1094,35 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
     }
 
     /// Utility function for *non*-scope code to build their own drops
+    /// Force a drop at this point in the MIR by creating a new block.
+    /// Should only be used for DropAndReplace.
+    /// Note that this does *not* insert the assignement in the MIR,
+    /// that should be added in the newly returned block by the caller.
     pub(crate) fn build_drop_and_replace(
         &mut self,
         block: BasicBlock,
         span: Span,
         place: Place<'tcx>,
-        value: Operand<'tcx>,
+        value: Rvalue<'tcx>,
     ) -> BlockAnd<()> {
         let source_info = self.source_info(span);
         let next_target = self.cfg.start_new_block();
 
+        let assign = self.cfg.start_new_cleanup_block();
+        self.cfg.push_assign(assign, source_info, place, value.clone());
+        self.cfg.terminate(assign, source_info, TerminatorKind::Goto { target: block });
+
         self.cfg.terminate(
             block,
             source_info,
-            TerminatorKind::DropAndReplace { place, value, target: next_target, unwind: None },
+            TerminatorKind::Drop {
+                place,
+                target: next_target,
+                unwind: Some(assign),
+                is_replace: true,
+            },
         );
+
         self.diverge_from(block);
```

Comment:
> I think that diverge_from will place block in a list and fix the terminator of assign later. Kind of confusing.

Yes, I initially considered the suggestion in the last paragraph of your comment but wanted to avoid unnecessary allocations. I guess this option can be made clearer with an obviously dummy target for the goto and a comment.

>Also, I presume that the goal of the diverge_from mechanism is to avoid creating unwind blocks when we are in panic=abort mode.

Good catch, I need to check, I don't remember something explicit about the panic strategy during mir build though, maybe it's shadowed by some other logic

Reply:
Why would this create more allocations? I haven't read into the code, but I imagine there is a `Vec<DivergeFrom>` stored *somewhere*, isn't it just going to make `DivergeFrom` a bit larger (because it now stores a `Option<(Place, Rvalue)>` pair or something like that)

Reply:
(That said, that might make the data structure significantly bigger.)

Reply:
> Also, I presume that the goal of the diverge_from mechanism is to avoid creating unwind blocks when we are in panic=abort mode.

after a quick investigation (MIR dump of the nightly compiler) it looks like unwind/cleanup blocks are always created and later removed by the `AbortUnwindingCalls` pass.

I changed the terminator to `TerminatorKind::Unreachable` and added a comment to make it clearer that it's only temporary

Reply:
Followed @oli-obk suggestion and added the terminator later